File: samples/CameraAccess/CameraAccess/AX022/Ax022GatewayBridge.swift
import Foundation

@MainActor
final class Ax022GatewayBridge: ObservableObject, ToolBridge {
  @Published var lastToolCallStatus: ToolCallStatus = .idle
  @Published var connectionState: OpenClawConnectionState = .notConfigured

  private let session: URLSession

  init() {
    let config = URLSessionConfiguration.default
    config.timeoutIntervalForRequest = 120
    self.session = URLSession(configuration: config)
  }

  func checkConnection() async {
    guard SettingsManager.shared.isAx022Configured else {
      connectionState = .notConfigured
      return
    }
    connectionState = .checking
    guard let url = URL(string: SettingsManager.shared.ax022GatewayURL.trimmingCharacters(in: CharacterSet(charactersIn: "/")) + "/v1/capabilities") else {
      connectionState = .unreachable("Invalid AX-022 URL")
      return
    }
    var request = URLRequest(url: url)
    request.httpMethod = "GET"
    request.setValue("Bearer \(SettingsManager.shared.ax022SessionToken)", forHTTPHeaderField: "Authorization")
    do {
      let (_, response) = try await session.data(for: request)
      if let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) {
        connectionState = .connected
      } else {
        let code = (response as? HTTPURLResponse)?.statusCode ?? 0
        connectionState = .unreachable("AX-022 HTTP \(code)")
      }
    } catch {
      connectionState = .unreachable(error.localizedDescription)
    }
  }

  func resetSession() {
    // The scoped AX-022 bearer token identifies the wearable session server-side.
  }

  func delegateTask(task: String, toolName: String) async -> ToolResult {
    lastToolCallStatus = .executing(toolName)
    let base = SettingsManager.shared.ax022GatewayURL.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
    guard let url = URL(string: base + "/v1/intent") else {
      lastToolCallStatus = .failed(toolName, "Invalid AX-022 URL")
      return .failure("Invalid AX-022 gateway URL")
    }

    var request = URLRequest(url: url)
    request.httpMethod = "POST"
    request.setValue("Bearer \(SettingsManager.shared.ax022SessionToken)", forHTTPHeaderField: "Authorization")
    request.setValue("application/json", forHTTPHeaderField: "Content-Type")

    do {
      request.httpBody = try JSONSerialization.data(withJSONObject: ["message": task])
      let (data, response) = try await session.data(for: request)
      let code = (response as? HTTPURLResponse)?.statusCode ?? 0
      guard (200...299).contains(code) else {
        lastToolCallStatus = .failed(toolName, "HTTP \(code)")
        return .failure("AX-022 returned HTTP \(code)")
      }

      if let payload = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
         let result = payload["result"] {
        let spoken: String
        if let resultDict = result as? [String: Any] {
          spoken = (resultDict["response"] as? String)
            ?? (resultDict["text"] as? String)
            ?? String(data: try JSONSerialization.data(withJSONObject: resultDict), encoding: .utf8)
            ?? "Done"
        } else if let resultString = result as? String {
          spoken = resultString
        } else {
          spoken = String(describing: result)
        }
        lastToolCallStatus = .completed(toolName)
        return .success(spoken)
      }

      let raw = String(data: data, encoding: .utf8) ?? "Done"
      lastToolCallStatus = .completed(toolName)
      return .success(raw)
    } catch {
      lastToolCallStatus = .failed(toolName, error.localizedDescription)
      return .failure("AX-022 agent error: \(error.localizedDescription)")
    }
  }
}

File: samples/CameraAccessAndroid/app/src/main/java/com/meta/wearable/dat/externalsampleapps/cameraaccess/ax022/Ax022GatewayBridge.kt
package com.meta.wearable.dat.externalsampleapps.cameraaccess.ax022

import android.util.Log
import com.meta.wearable.dat.externalsampleapps.cameraaccess.openclaw.OpenClawConnectionState
import com.meta.wearable.dat.externalsampleapps.cameraaccess.openclaw.ToolBridge
import com.meta.wearable.dat.externalsampleapps.cameraaccess.openclaw.ToolCallStatus
import com.meta.wearable.dat.externalsampleapps.cameraaccess.openclaw.ToolResult
import com.meta.wearable.dat.externalsampleapps.cameraaccess.settings.SettingsManager
import java.util.concurrent.TimeUnit
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject

/**
 * Thin Android client for the server-side AX-022 gateway.
 *
 * The phone stores only the scoped AX-022 wearable session token. Tenant agent
 * credentials such as MAXX_API_KEY remain on the gateway host.
 */
class Ax022GatewayBridge : ToolBridge {
    companion object {
        private const val TAG = "Ax022GatewayBridge"
    }

    private val _lastToolCallStatus = MutableStateFlow<ToolCallStatus>(ToolCallStatus.Idle)
    override val lastToolCallStatus: StateFlow<ToolCallStatus> = _lastToolCallStatus.asStateFlow()

    private val _connectionState = MutableStateFlow<OpenClawConnectionState>(OpenClawConnectionState.NotConfigured)
    override val connectionState: StateFlow<OpenClawConnectionState> = _connectionState.asStateFlow()

    private val client = OkHttpClient.Builder()
        .readTimeout(120, TimeUnit.SECONDS)
        .connectTimeout(10, TimeUnit.SECONDS)
        .build()

    override fun setToolCallStatus(status: ToolCallStatus) {
        _lastToolCallStatus.value = status
    }

    override suspend fun checkConnection() = withContext(Dispatchers.IO) {
        if (!SettingsManager.isAx022Configured) {
            _connectionState.value = OpenClawConnectionState.NotConfigured
            return@withContext
        }

        _connectionState.value = OpenClawConnectionState.Checking
        try {
            val request = Request.Builder()
                .url("${SettingsManager.ax022GatewayUrl.trimEnd('/')}/v1/capabilities")
                .get()
                .addHeader("Authorization", "Bearer ${SettingsManager.ax022SessionToken}")
                .build()
            val response = client.newCall(request).execute()
            response.close()
            _connectionState.value = if (response.isSuccessful) {
                OpenClawConnectionState.Connected
            } else {
                OpenClawConnectionState.Unreachable("AX-022 HTTP ${response.code}")
            }
        } catch (e: Exception) {
            _connectionState.value = OpenClawConnectionState.Unreachable(e.message ?: "Unknown error")
            Log.d(TAG, "AX-022 gateway unreachable: ${e.message}")
        }
    }

    override fun resetSession() {
        // Server-side AX-022 session identity is represented by the scoped bearer token.
    }

    override suspend fun delegateTask(task: String, toolName: String): ToolResult = withContext(Dispatchers.IO) {
        if (!SettingsManager.isAx022Configured) {
            return@withContext ToolResult.Failure("AX-022 gateway is not configured")
        }

        _lastToolCallStatus.value = ToolCallStatus.Executing(toolName)
        try {
            val body = JSONObject().put("message", task)
            val request = Request.Builder()
                .url("${SettingsManager.ax022GatewayUrl.trimEnd('/')}/v1/intent")
                .post(body.toString().toRequestBody("application/json".toMediaType()))
                .addHeader("Authorization", "Bearer ${SettingsManager.ax022SessionToken}")
                .addHeader("Content-Type", "application/json")
                .build()

            val response = client.newCall(request).execute()
            val responseText = response.body?.string().orEmpty()
            val status = response.code
            response.close()

            if (status !in 200..299) {
                _lastToolCallStatus.value = ToolCallStatus.Failed(toolName, "HTTP $status")
                return@withContext ToolResult.Failure("AX-022 returned HTTP $status")
            }

            val payload = JSONObject(responseText)
            val result = payload.opt("result")
            val spoken = when (result) {
                is JSONObject -> result.optString("response").ifBlank {
                    result.optString("text").ifBlank { result.toString() }
                }
                null -> responseText
                else -> result.toString()
            }

            _lastToolCallStatus.value = ToolCallStatus.Completed(toolName)
            ToolResult.Success(spoken)
        } catch (e: Exception) {
            Log.e(TAG, "AX-022 agent error: ${e.message}")
            _lastToolCallStatus.value = ToolCallStatus.Failed(toolName, e.message ?: "Unknown")
            ToolResult.Failure("AX-022 agent error: ${e.message}")
        }
    }
}
